### DISCUSSION TITLE
Error if `dns_servers` fails and `dns_cache` per request.

### DISCUSSION THREAD

**Discussion by kathampy** [sagebind/isahc#103] [General]
It's hard to tell if configured DNS servers are being used at all and if `libcurl` was built with `c-ares`. Custom DNS servers don't seem to be working on macOS, with no error.

I'm also forced to use the default HTTP client and cannot set the DNS cache duration. I want to disable it for every request on the default client. This is another thing that would benefit from an `isolate()` that sets a new share handle on the request

**Comment by sagebind** [comment]
> It's hard to tell is configured DNS servers are being used at all and if libcurl was built with C-ARES.

It should be reflected if it is using c-ares in `isahc::version()`. Truthfully it is rare to see libcurl compiled with c-ares, so `dns_servers` will rarely have any effect. If I remember correctly we write a warning to the logs when calling this if it will not have an effect, but I could be wrong.

DNS cache has not been released yet as I'm not ready to commit to the current design yet.

**Comment by sagebind** [comment]
As far as erroring goes, it would be a backwards-incompatible change to begin returning an error from `dns_servers()`. It looks like it might be possible to detect if c-ares is enabled at runtime, if that is something we want to do: https://docs.rs/curl/0.4.25/curl/struct.Version.html#method.ares_version

**Comment by kathampy** [comment]
For those interested in using the system `libcurl` or their own build with `c-ares`:
```
# Remove the default `static-curl` feature in `isahc` so that `libcurl`
# built with `c-ares` and `openssl` is used. Override the system `libcurl` with
# `RUSTFLAGS=-L/usr/local/opt/curl-openssl/lib cargo run`.
isahc = {version = "0.7.4", default-features = false, features = ["cookies", "http2"]}
```

**Comment by kathampy** [comment]
@sagebind Custom DNS servers per request is complemented by disabling DNS caching for that request. My benchmarking tool needs this capability.
DNS caching should be allowed only for redirects within that request, but not for other requests. Disabling DNS caching for the request would would cause two DNS requests for HTTP to HTTPS redirects as the connection wouldn't be reused. Isolating the request with a share handle seems to be the only way again.

**Comment by sagebind** [comment]
Don't worry, its on my todo list! I'll work on it as soon as I have time.
